docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,76 @@
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. There is a default that allows deriving keys for buckets when no bucket specific secret is provided.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created in the future.
+
+## Design
+
+Uplink always stores derived keys, never the input material.
+
+The satellite will add a new endpoint, `GetSaltEntropy` that uses the api key to find the project, and optionally takes a bucket name. It is important that the result is stable, even if the bucket is renamed. For example the result of `GetSaltEntropy("bucket1")` should be the same as `GetSaltEntropy("bucket2")` even after `"bucket1"` is renamed to `"bucket2"`. Note that there are still significant challenges to bucket renaming, and we don't support it right now.
+
+When deriving a root key from an api key, password, optional bucket name and optional encrypted path, the following algorithm will be used
+
+```
+saltEntropy = GetSaltEntropy(bucketName or None)
+mixedSalt   = hmac(hash:sha256, secret:password, data:saltEntropy)
+pathSalt    = hmac(hash:sha256, secret:mixedSalt, data:encryptedPath or "")
+rootKey     = argon2id(salt:salt, password:password)
```

Comment:
`salt:salt` here is presumably `salt:pathSalt` or `secret:pathSalt`? I don't know what this colon syntax is.

Reply:
I just now realized that while I was trying to do python kwargs, it uses `=` instead of `:`. It should be `pathSalt`.